get_data_ready.yml

```diff
@@ -0,0 +1,98 @@
+target_default: all
+
+packages:
+  - dplyr
+  - readxl
+  - scipiper
+  - tidyr
+  - stringr
+  - data.table
+  - yaml
+  - dataRetrieval
+  - EGRET
+  - lubridate
+  - knitr
+
+sources:
+  - 1_get_data_ready/src/get_raw_data.R
+  - 1_get_data_ready/src/clean_sample_data.R
+  - 1_get_data_ready/src/clean_censored_data.R
+  - 1_get_data_ready/src/summarize_sites.R
+  - 1_get_data_ready/src/summarize_flow.R
+  - 1_get_data_ready/src/get_flow.R
+  - 1_get_data_ready/src/adjust_discharge_data.R
+  - 1_get_data_ready/src/merge_sample_flow.R
+  - 1_get_data_ready/src/render.R
+
+file_extensions:
+  - ind
+
+targets:
+  all:
+    depends: 
+    - lib/cfg/gd_config.yml
+    - 1_get_data_ready/doc/data_checks.pdf.ind
+    - 1_get_data_ready/doc/diagram.html.ind
+    - 1_get_data_ready/out/merged_data.rds.ind
+
+  lib/cfg/gd_config.yml:
+    command: gd_config(config_file=target_name, folder=I("1DGJJ1_tc1_92v0uOwL1bqFfPPyqk-nyZ"))
+  
+  1_get_data_ready/out:
+    command: dir.create(target_name, recursive=I(TRUE), showWarnings=I(FALSE))
+
+  1_get_data_ready/doc:
+    command: dir.create(target_name, recursive=I(TRUE), showWarnings=I(FALSE))
+
+  1_get_data_ready/out/USGS_WQ_DATA_02-16.xlsx.ind:
+    command: gd_confirm_posted(ind_file = target_name)    
+
+  1_get_data_ready/out/SampleGanttCharts_wRanks.xlsx.ind:
+    command: gd_confirm_posted(ind_file = target_name)
+    
+  raw_sample:
+    command: get_sample("1_get_data_ready/out/USGS_WQ_DATA_02-16.xlsx.ind")
+    
+  raw_site:
+    command: get_sites("1_get_data_ready/out/SampleGanttCharts_wRanks.xlsx.ind")
+
+  clean_sample: 
+    command: clean_sample_data(raw_sample)
+    
+  sample_data:
+    command: censored_data(clean_sample)
+    
+  summary_sites: 
+    command: summarize_sites(sample_data, raw_site) 
+    
+  summary_flow:
+    command: summarize_flow(summary_sites)
+    
+  raw_flow:
+    command: get_flow(summary_flow)
+
+  adjusted_discharge: 
+    command: adjust_discharge_data(raw_flow, summary_sites)
+  
+  merged_flow: 
+    command: merge_sample_flow(sample_data, summary_sites, adjusted_discharge, I('1_get_data_ready/out'))
+  
+  1_get_data_ready/doc/data_checks.pdf:
+    command: plot_eLists(merged_flow, '1_get_data_ready/out', target_name)
+
+  1_get_data_ready/doc/data_checks.pdf.ind:
+    command: gd_put(target_name, '1_get_data_ready/doc/data_checks.pdf')
```

Comment:
So if Steve never tries to run remake, this is great. and if he's content to rebuild data_checks.pdf and all its dependencies, this is great. But if you're aiming for Steve (or any collaborator) to be able to clone the repo and run `scmake` with the result that his computer only pulls down files from Google Drive and doesn't build things locally, this won't work.

What will happen instead, if you build the whole repo and Steve creates a fresh clone from GitHub and runs `scmake`, is this: remake goes to build `1_get_data_ready/doc/data_checks.pdf.ind`, notices that it depends on `1_get_data_ready/doc/data_checks.pdf`, notices that THAT depends on `merged_flow`, and on back up the chain. Because GitHub doesn't contain R objects or `1_get_data_ready/doc/data_checks.pdf`, remake will decide that it needs to build them all to determine whether `1_get_data_ready/doc/data_checks.pdf.ind` is up to date. Even though we have a build/status file in GitHub that claims this is the case. Result: Steve's computer builds everything.

Again, having Steve's computer build everything might be just fine if this is all pretty quick. But if you're aiming for the shared-cache model where Steve's computer only downloads the final files, read on.

1. To make it so that `scmake('1_get_data_ready/doc/data_checks.pdf')` causes a call to `gd_get('1_get_data_ready/doc/data_checks.pdf.ind')` instead of causing everything to be built again, make a new target after `1_get_data_ready/doc/data_checks.pdf.ind`:

```
1_get_data_ready/doc/data_checks.pdf:
  command: gd_get('1_get_data_ready/doc/data_checks.pdf.ind')
```

2. This introduces duplicate targets (two `1_get_data_ready/doc/data_checks.pdf`s). So you need to rename the first one if you want the second one to exist. I recommend making the target be an `ind` file within a `tmp` folder, either in tmpdir() (not shown) or your repo (shown). plot_eLists should be responsible for creating the .pdf file behind the scenes (the pdf must be exactly the result of `as_data_file` applied to this target name, i.e., `1_get_data_ready/tmp/data_checks.pdf`). `plot_eLists` can't tell remake about the data file because annoying things (redundant builds) would happen. So instead it should return the .ind version of the file.
```
1_get_data_ready/tmp/data_checks.pdf.ind:
  command: plot_eLists(merged_flow, '1_get_data_ready/out', target_name)
```
Then in your next remake recipe you can refer to that tmp/data_checks.pdf.ind file in the recipe for gd_put:
```
1_get_data_ready/doc/data_checks.pdf.ind:
  command: gd_put(remote_ind=target_name, local_source='1_get_data_ready/tmp/data_checks.pdf.ind')
```

3. With two targets for data_checks.pdf, your own computer will now build, upload, _and then download_ the pdf, which is wasteful. You can sidestep the need to download a file that's already on your computer by making sure the `mock_get` argument is set. Here, I think you could use either `'copy'` or `'move'`. `copy` is the default, so you don't technically need to change anything from what's above, but here's the explicit version (remember the `I()`):
```
1_get_data_ready/doc/data_checks.pdf.ind:
  command: gd_put(remote_ind=target_name, local_source='1_get_data_ready/tmp/data_checks.pdf',
                                 mock_get=I('copy'))
```

I put some examples of all this in the help doc for `?gd_put`.

4. Hmm. Even after all that, this particular example isn't going to come out squeaky clean, because every computer that builds `1_get_data_ready/doc/data_checks.pdf` will also want to build all the dependencies of `1_get_data_ready/tmp/data_checks.pdf.ind`. If you don't want this to happen, you could (a) make the upstream objects into files, and only include their .ind versions in the remake file, like we just did for 1_get_data_ready/tmp/data_checks.pdf.ind, or (b) break the dependency chain that remake sees in this yml, which you could do by (b1) using I() in the recipe for `1_get_data_ready/doc/data_checks.pdf`, as in
```
1_get_data_ready/doc/data_checks.pdf:
  command: gd_get(I('1_get_data_ready/doc/data_checks.pdf.ind'))
```
or (b2) putting the recipe for `1_get_data_ready/doc/data_checks.pdf` in a separate remake.yml that does not `include` this remake.yml. I suggest this option in a comment below, because it seems like the shortest path from what you have to what might be nice. However, option (a) is the only one that preserves a complete dependency chain throughout your build system, whereas options (b1) and (b2) would both require manual care if your repo ever lacks `1_get_data_ready/doc/data_checks.pdf.ind` or if your Drive folder ever lacks `1_get_data_ready/doc/data_checks.pdf`.

Reply:
I'm still reading, but I wanted to state that yes...I rebuilt this with the intention that that Steve/Sam won't re-run it unless they really want to troubleshoot what was going on with the data prep phase (ie...this is the "sneaky yml" that we were discussing). From here I've sent the final eList_list to Condor and have a bunch of data results to play with. My plan is that the next remake.yml file will be the one we'd want the whole team to play with. The initial targets for that remake recipe would be the indicator files of the data checks I think.

Reply:
Fine with me. I'm hoping the thought process laid out still gives you tools for thinking about future projects with different requirements. Note that for the next remake.yml to be something that everybody can run without re-running this one, the next yml probably shouldn't `include` this one.

Reply:
Definitely